File: app/src/main/java/knf/kuma/database/dao/ChaptersDAO.java
package knf.kuma.database.dao;

import java.util.List;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.TypeConverters;
import knf.kuma.database.BaseConverter;
import knf.kuma.pojos.AnimeObject;

@Dao
@TypeConverters({BaseConverter.class})
public interface ChaptersDAO {

    @Query("SELECT count(*) FROM animechapter")
    int init();

    @Query("SELECT * FROM animechapter")
    List<AnimeObject.WebInfo.AnimeChapter> getAll();

    @Query("SELECT * FROM animechapter WHERE eid = :eid LIMIT 1")
    LiveData<AnimeObject.WebInfo.AnimeChapter> chapterSeen(String eid);

    @Query("SELECT count(*) FROM animechapter WHERE eid = :eid")
    Boolean chapterIsSeen(String eid);

    @Query("SELECT * FROM animechapter WHERE eid IN (:eids) ORDER BY eid DESC LIMIT 1")
    AnimeObject.WebInfo.AnimeChapter getLast(List<String> eids);

    @Query("SELECT * FROM animechapter WHERE aid LIKE :aid ORDER BY `key` DESC LIMIT 1")
    AnimeObject.WebInfo.AnimeChapter getLastByAid(String aid);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void addChapter(AnimeObject.WebInfo.AnimeChapter chapter);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void addAll(List<AnimeObject.WebInfo.AnimeChapter> list);

    @Delete
    void deleteChapter(AnimeObject.WebInfo.AnimeChapter chapter);

    @Query("DELETE FROM animechapter")
    void clear();
}

File: app/src/main/java/knf/kuma/database/dao/ChaptersDAO.kt
package knf.kuma.database.dao

import androidx.lifecycle.LiveData
import androidx.room.*
import knf.kuma.database.BaseConverter
import knf.kuma.pojos.AnimeObject

@Dao
@TypeConverters(BaseConverter::class)
interface ChaptersDAO {

    @get:Query("SELECT * FROM animechapter")
    val all: MutableList<AnimeObject.WebInfo.AnimeChapter>

    @get:Query("SELECT count(*) FROM animechapter")
    val countLive: LiveData<Int>

    @get:Query("SELECT count(*) FROM animechapter")
    val count: Int

    @Query("SELECT * FROM animechapter WHERE eid = :eid LIMIT 1")
    fun chapterSeen(eid: String): LiveData<AnimeObject.WebInfo.AnimeChapter>

    @Query("SELECT count(*) FROM animechapter WHERE eid = :eid")
    fun chapterIsSeen(eid: String): Boolean

    @Query("SELECT * FROM animechapter WHERE eid IN (:eids) ORDER BY eid+0 DESC LIMIT 1")
    fun getLast(eids: MutableList<String>): AnimeObject.WebInfo.AnimeChapter?

    @Query("SELECT * FROM animechapter WHERE aid = :aid ORDER BY `key` DESC LIMIT 1")
    fun getLastByAid(aid: String): AnimeObject.WebInfo.AnimeChapter?

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun addChapter(chapter: AnimeObject.WebInfo.AnimeChapter)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun addAll(list: List<AnimeObject.WebInfo.AnimeChapter>)

    @Delete
    fun deleteChapter(chapter: AnimeObject.WebInfo.AnimeChapter)

    @Query("DELETE FROM animechapter")
    fun clear()
}
